sn_transfers/src/payment_proof/mod.rs

```diff
@@ -0,0 +1,252 @@
+use hasher::Sha256Hasher;
+
+use sn_protocol::{
+    messages::{Hash, PaymentProof},
+    NetworkAddress,
+};
+
+use merkletree::{
+    hash::Algorithm,
+    merkle::{next_pow2, MerkleTree},
+    proof::Proof,
+    store::VecStore,
+};
+use std::collections::BTreeMap;
+use typenum::{UInt, UTerm, B0, B1};
+use xor_name::XorName;
+
+// Data type of each of the nodes in the binary Merkle-tree we build payment proofs with
+type MerkleTreeNodesType = [u8; 32];
+
+// We use a binary Merkle-tree to build payment proofs
+type BinaryMerkletreeProofType = Proof<MerkleTreeNodesType, UInt<UInt<UTerm, B1>, B0>>;
+
+/// Map from content address name to its corresponding PaymentProof
+pub type PaymentProofsMap = BTreeMap<[u8; 32], PaymentProof>;
+
+/// Build a Merkletree to generate the PaymentProofs for each of the content addresses provided
+// TODO: provide fix against https://en.wikipedia.org/wiki/Preimage_attack ?
+pub fn build_payment_proofs<'a>(
+    content_addrs: impl Iterator<Item = &'a NetworkAddress>,
+) -> Result<(Hash, PaymentProofsMap)> {
+    // Let's build the Merkle-tree from list of addresses needed to generate the payment proofs
+    // TODO: bail out if the list is empty
+    let mut addrs: Vec<_> = content_addrs
+        .map(|addr| {
+            let mut arr = MerkleTreeNodesType::default();
+            // TODO: check the length of addr is 32 so we don't (unexpectedly) miss bytes...?
+            arr.copy_from_slice(&addr.as_bytes());
+            arr
+        })
+        .collect();
+
+    // Merkletree requires the number of leaves to be a power of 2, and at least 2 leaves.
+    let num_of_leaves = usize::max(2, next_pow2(addrs.len()));
+    let num_of_addrs = addrs.len();
+    for _ in num_of_addrs..num_of_leaves {
+        // fill it up with blank value leaves
+        addrs.push(MerkleTreeNodesType::default());
+    }
+
+    let merkletree = MerkleTree::<MerkleTreeNodesType, Sha256Hasher, VecStore<_>>::new(
+        addrs.clone().into_iter(),
+    )
+    .map_err(|err| Error::ProofTree(err.to_string()))?;
+
+    // The reason hash is set to be the root of the merkle-tree of chunks to pay for
+    let reason_hash = merkletree.root().into();
+
+    let mut payment_proofs = BTreeMap::new();
+    for (index, addr) in addrs.into_iter().take(num_of_addrs).enumerate() {
+        let proof = merkletree
+            .gen_proof(index)
+            .map_err(|err| Error::GenAuditTrail {
+                index,
+                reason: err.to_string(),
+            })?;
+
+        payment_proofs.insert(
+            addr,
+            PaymentProof {
+                reason_hash,
+                lemma: proof.lemma().to_vec(),
+                path: proof.path().to_vec(),
+            },
+        );
+    }
+
+    Ok((reason_hash, payment_proofs))
+}
+
+/// Verify if the payment proof is valid and contains a valid audit trail for the given xorname
+pub fn validate_payment_proof(addr_name: XorName, payment: &PaymentProof) -> Result<()> {
```

Comment:
I'm wondering if there's good literature explaining the merkle tree and chunk addressing we can link to here. Would be good to include in the comments if you've come across anything decent @bochaco ?

Reply:
I'll add a paragraph with some info and a link.

Reply:
Lovely stuff 💪